File: src/main/java/nl/rabobank/kotlinmovement/recipes/controller/RecipesControllerAdvice.java
package nl.rabobank.kotlinmovement.recipes.controller;

import nl.rabobank.kotlinmovement.recipes.model.RecipesErrorResponse;
import nl.rabobank.kotlinmovement.recipes.service.ResourceNotFoundException;
import nl.rabobank.kotlinmovement.recipes.util.ErrorMessageUtil;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;

import java.util.List;

@ControllerAdvice()
public class RecipesControllerAdvice {

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<RecipesErrorResponse> validationException(MethodArgumentNotValidException e) {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
                new RecipesErrorResponse(
                        ErrorMessageUtil.toErrorMessage(e, "Incorrect fields:", ".")
                )
        );
    }

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<RecipesErrorResponse> resourceNotFoundException(ResourceNotFoundException e) {
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(
                new RecipesErrorResponse(
                        ErrorMessageUtil.toErrorMessage(List.of(e.getMessage()), null, null)
                )

        );
    }
    @ExceptionHandler(Exception.class)
    public ResponseEntity<RecipesErrorResponse> defaultExceptionHandler(Exception exception) {
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(
                new RecipesErrorResponse(
                        ErrorMessageUtil.DEFAULT_ERROR_MESSAGE
                )
        );
    }
}

File: non-blocking-complete/src/main/kotlin/nl/alimeshkat/recipes/controller/RecipesControllerAdvice.kt
package nl.alimeshkat.recipes.controller

import nl.alimeshkat.recipes.model.RecipesErrorResponse
import nl.alimeshkat.recipes.service.ResourceNotFoundException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.support.WebExchangeBindException

@ControllerAdvice
class RecipesControllerAdvice {

    @ExceptionHandler(WebExchangeBindException::class)
    fun validationException(e: WebExchangeBindException): ResponseEntity<RecipesErrorResponse> {
        log.error("error", e)
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
            ErrorMessageMapper.toErrorMessage(e, "Incorrect fields:", ".")
        )
    }

    @ExceptionHandler(ResourceNotFoundException::class)
    fun resourceNotFoundException(e: ResourceNotFoundException): ResponseEntity<RecipesErrorResponse> {
        log.error("error", e)
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(
            ErrorMessageMapper.toErrorMessage(listOf(e.message))
        )
    }

    @ExceptionHandler(Exception::class)
    fun defaultExceptionHandler(e: Exception?): ResponseEntity<RecipesErrorResponse> {
        log.error("error", e)
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(
            ErrorMessageMapper.DEFAULT_ERROR_MESSAGE
        )
    }

    companion object {
        val log: Logger = LoggerFactory.getLogger(RecipesControllerAdvice::class.java)
    }
}
